include/powsybl/iidm/ExtensionProviders.hxx

```diff
@@ -46,10 +45,72 @@ const T& ExtensionProviders<T, Dummy>::findProviderOrThrowException(const std::s
 }
 
 template <typename T, typename Dummy>
-void ExtensionProviders<T, Dummy>::registerExtension(const std::string& name, std::unique_ptr<T>&& provider) {
-    getInstance().m_providers.emplace(std::make_pair(name, std::move(provider)));
+std::vector<std::string> ExtensionProviders<T, Dummy>::getFiles(const std::string& directory, const boost::regex& file) {
+    std::vector<std::string> libs;
+    if (boost::filesystem::exists(directory) && boost::filesystem::is_directory(directory)) {
+        for (const auto& it : boost::filesystem::directory_iterator(directory)) {
+            if (boost::filesystem::is_regular_file(it) && boost::regex_match(it.path().filename().c_str(), file)) {
+                libs.push_back((directory / it.path().filename()).string());
+            }
+        }
+    }
+
+    return libs;
+}
+
+template <typename T, typename Dummy>
+ExtensionProviders<T, Dummy>& ExtensionProviders<T, Dummy>::getInstance() {
+    static ExtensionProviders<T, Dummy> s_instance;
+    return s_instance;
 }
 
+template <typename T, typename Dummy>
+void ExtensionProviders<T, Dummy>::loadExtensions(const std::string& path, const boost::regex& files) {
+    if (!boost::filesystem::exists(path)) {
+        throw PowsyblException(logging::format("Path %1% does not exist", path));
+    }
+
+    if (!boost::filesystem::is_directory(path)) {
+        throw PowsyblException(logging::format("Path %1% is not a directory", path));
+    }
+
+    logging::Logger& logger = logging::LoggerFactory::getLogger<ExtensionProviders>();
+    const std::vector<std::string>& libFiles = getFiles(path, files);
+    for (const std::string& libFile : libFiles) {
+        try {
+            loadLibrary(libFile);
+        } catch (const boost::system::system_error& err) {
+            logger.info(logging::format("Unable to load file %1%: %2%", libFile, err.what()));
+        }
+    }
+}
+
+template <typename T, typename Dummy>
+void ExtensionProviders<T, Dummy>::loadLibrary(const std::string& library) {
+    logging::Logger& logger = logging::LoggerFactory::getLogger<ExtensionProviders>();
+    boost::dll::shared_library lib(library);
+
+    const boost::filesystem::path& libraryPath = boost::filesystem::canonical(lib.location());
+    if (m_loadedLibraries.find(libraryPath) == m_loadedLibraries.end()) {
+        if (lib.has("createSerializers")) {
+            const auto& createSerializers = boost::dll::import_alias<std::set<std::unique_ptr<T>>()>(lib, "createSerializers");
+            for (auto& it : createSerializers()) {
+                const std::string& extensionName = it->getExtensionName();
+                registerExtension(extensionName, std::move(const_cast<std::unique_ptr<T>&>(it)));
```

Comment:
I know, this const_cast is ugly but I could not find out what's wrong without :/